include/zephyr/bluetooth/audio/audio.h

```diff
@@ -1784,6 +1784,58 @@ int bt_audio_unicast_server_register_cb(const struct bt_audio_unicast_server_cb
  */
 int bt_audio_unicast_server_unregister_cb(const struct bt_audio_unicast_server_cb *cb);
 
+/**
+ * @brief Structure holding information of audio stream endpoint
+ */
+struct bt_audio_ep_info {
+	/** The ID of the ASE */
+	uint8_t id;
+	/** The state of the endpoint */
+	uint8_t state;
+	/** Capabilities type */
+	enum bt_audio_dir dir;
+};
+
+/**
+ * @brief Return structure holding information of audio stream endpoint
+ *
+ * @param ep   The audio stream endpoint object.
+ * @param info The structure object to be filled with the info.
+ */
+int bt_audio_ep_get_info(struct bt_audio_ep *ep, struct bt_audio_ep_info *info);
+
+/** @typedef bt_audio_ase_func_t
+ *  @brief The callback function called for each ASE.
+ *
+ *  @param info The structure object with ASE info.
+ *  @param user_data Data to pass to the function.
+ */
+typedef void (*bt_audio_ase_func_t)(const struct bt_audio_ep_info *info,
+				    void *user_data);
+
+/**
+ * @brief Iterate through all ASE of the given connection.
+ *
+ * @param conn Connection object
+ * @param func Function to call for each ASE.
+ * @param user_data Data to pass to the callback function.
+ */
+void bt_audio_unicast_server_foreach_ase(struct bt_conn *conn,
+		bt_audio_ase_func_t func, void *user_data);
+
+/**
+ * @brief Init and configure ASE.
+ *
+ *  @param conn Connection object
+ *  @param ase_id ID of the ASE being configured
+ *  @param stream Configured stream object to be attached to the ASE
+ *  @param codec Codec configuration
+ *  @param qos_pref Audio Stream Quality of Service Preference
+ */
+int bt_audio_unicast_server_config_ase(struct bt_conn *conn, uint8_t ase_id,
```

Comment:
Why do we need this, rather than using `bt_audio_stream_config`? 

Given the `ep_info`, we can now convert `ase_id` to `* ep`. 

One thing that does come to mind for me, is that the `qos_pref` would be missing, but we may get around by calling the `configure` callback with some way of indicating that this is a local call (which arguably can be done by comparing the `stream` or `ep` pointer.

Reply:
The point is, I'd rather avoid adding a new function for this if we can, but you may have thought of some reason why we can't do that :)

Reply:
I've consulted the issue with @MariuszSkamra today and we came to the conclusion that ase object does not exist until remote client triggers Codec Config.

So if the server needs to initiate the Codec Config before the client, an ase has to be created/inited first. Earlier I used `ase_get()` in `bt_unicast_server_foreach_ase()`:
```
	for (size_t i = 1; i <= ASE_COUNT; i++) {
		ase = ase_get(ascs, i);
```
 which silently created an ase for me. But after refactor of the loop:
```
	SYS_SLIST_FOR_EACH_NODE(&ascs->ases, ase_node) {
		ase = CONTAINER_OF(ase_node, struct bt_ascs_ase, node);
```
the `ascs->ases` list is empty, so does not work for me anymore.

So there are two options I can see:
- a function similar to `bt_audio_unicast_server_config_ase` which openly creates and configures the ase,
- or a public version of ase_get(), e.g. `struct bt_audio_ep * bt_audio_unicast_server_init_ase(struct bt_conn *conn, uint8_t ase_id);`, which openly creates and returns the ase, so we have the needed `ep` for `bt_audio_stream_config`.

Reply:
`ase_get` is really a poor name for something that allocates a new ASE :( 

But yeah, I see the challenge. The `foreach` won't help us here, as the ASEs may not have been allocated for the ASCS instance. The modification to dynamically allocated ASEs made this a bit tricky :) 

Perhaps the best solution is to implement a `bt_audio_unicast_server_config_ase` as suggested. 

This should perhaps just take a `conn` and a `codec`, as the existing callback can be used to get the `stream` and `qos_pref`. 

The current suggested one can also work equally well if we omit the callback and then have `conn`, `stream`, `codec` and `qos_pref` as parameters. I'd like to avoid the `ase_id`, as the actual value of that is of no use to the application, and we have a few macros such as `ASE_DIR` and `ASE_UUID` that depend on ASE IDs being specific values. 

So I suggest to either have either
```c
int bt_audio_unicast_server_config_ase(struct bt_conn *conn, struct bt_codec *codec)
/* Depend on callback for the rest */
```

or 
```c
int bt_audio_unicast_server_config_ase(struct bt_conn *conn, struct bt_audio_stream *stream, struct bt_codec *codec, const struct bt_codec_qos_pref *qos_pref)
```

and then have the stack allocate the ASE ID.

Reply:
I've selected the second version. Now the server is able to create all ASEs on a connection event callback, just before the PTS Upper Tester enables notifications, so PTS won't receive any too early. When PTS finally requests from IUT to initiate Codec Config, the server can find proper stream by ASE id from its stream array and pass the stream to the
```C
int bt_audio_unicast_server_config_ase(struct bt_conn *conn, struct bt_audio_stream *stream, struct bt_codec *codec, const struct bt_codec_qos_pref *qos_pref);
```